Partner task submission documentation (#387)
redhat-appstudio/build-definitions: partners/README.md

Reviewer: Really pedantic, but should `yaml` be changed to `YAML`, `tekton` to `Tekton` and `openshift` to `OpenShift`?

Reviewer: Sorry, didn't notice that tkdchen suggested it already

Author: Added the change.

# Partner Tekton Task submission process

> Partners can submit their Tekton Task in [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository by sending a pull request.

Reviewer 2: https://github.com/redhat-appstudio/book/blob/main/ADR/0021-partner-tasks.md should be linked somewhere here.

Author: Added one extra line in this section referring the ADR `For more information, please refer the ADR`

build-definitions repository CI job will validate the PR against a [set of checks](#checks).

build-definitions repository maintainers needs to approve and merge the pull request before the Task can be used.

### How to submit a Task?

Create your Task YAML and put it inside a directory of structure `partners/<task_name>/<task_version>/<task_name>.yaml` and 
also include a `OWNERS` file inside the Task directory `partners/<task_name>/OWNERS`

For example, if your Task name is `my_task` and the version of the your Task is `0.1`, your Task YAML and OWNERS file location will be `partners/my_task/0.1/my_task.yaml` and `partners/my_task/OWNERS`

Send a pull request to the [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository containing your Task YAML and OWNERS file.

> ### How to wire the Task with the users pipeline?

Reviewer 2: Unfortunatelly this will not work for user. User will do this in user's repository, so it cannot point to  external. Currently we do not provide way how to use partner task.

Author: @Michkov Thanks for the input. For now I will remove this section. It would make sense when we have a way to integrate it in future.

Reviewer 2: or we can describe how to push the task into their bundle and how to consume.

# Partner Tekton Task submission process

Partners can submit their Tekton Task in [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository by sending a pull request.

build-definitions repository CI job will validate the PR against a [set of checks](#checks).

> build-definitions repository maintainers needs to approve and merge the pull request before the Task can be used.

Reviewer: ```suggestion
build-definitions repository maintainers need to approve and merge the pull request before the Task can be used.
```

Author: Done

> For more information, please refer the [ADR](https://github.com/redhat-appstudio/book/blob/main/ADR/0021-partner-tasks.md).

Reviewer: ```suggestion
For more information, please refer to the [ADR](https://github.com/redhat-appstudio/book/blob/main/ADR/0021-partner-tasks.md).
```

Author: Done

### How to submit a Task?

Create your Task YAML and put it inside a directory of structure `partners/<task_name>/<task_version>/<task_name>.yaml` and 

> also include a `OWNERS` file inside the Task directory `partners/<task_name>/OWNERS`

Reviewer: ```suggestion
also include an `OWNERS` file inside the Task directory `partners/<task_name>/OWNERS`
```

Author: Done

> For example, if your Task name is `my_task` and the version of your Task is `0.1`, your Task YAML and OWNERS file location will be `partners/my_task/0.1/my_task.yaml` and `partners/my_task/OWNERS`

Reviewer: ```suggestion
For example, if your Task name is `my_task` and the version of your Task is `0.1`, your Task YAML and OWNERS file locations will be `partners/my_task/0.1/my_task.yaml` and `partners/my_task/OWNERS`
```

Author: Its a single location.

Send a pull request to the [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository containing your Task YAML and OWNERS file.

### How to debug the CI failures in my submitted Task?

Check the logs of the `check-partner-tasks` Task in the `build-definitions-pull-request` PR check.
If you see the Task `yaml-lint-check` has failed, then your Task YAML contains yaml-lint errors.

### Checks

Task is validated against the following mandatory checks:

> * Task follows the correct directory structure and includes OWNERS file (This check is a prerequisites for other checks to be continued)

Reviewer: ```suggestion
* Task follows the correct directory structure and includes OWNERS file (This check is a prerequisite for other checks to be continued)
```

* Task is not using any privilege escalations
* Task schema is a valid one i.e. the Task resource can be applied in the Openshift cluster

## FAQs

### Can I validate the Task locally?

Yes, you can also validate your Task by running the script `hack/check-partner-tasks.sh` in your local clone. 
Make sure you have the `oc` client installed and access to an Openshift cluster.

> Login to the Openshift cluster first before running the script, otherwise Task schema validation will be ignored.

Reviewer: ```suggestion
Log in to the Openshift cluster first before running the script, otherwise Task schema validation will be ignored.
```

Author: Done.

### Can partner engineers revoke a Task?

Yes, they can also submit a pull request removing the Task directory from `partners` directory.

> Before sending the pull request make sure that revoking a Task will not be impacting any users pipeline.

Reviewer: If you mean "will not impact the pipelines of multiple users":
```suggestion
Before sending the pull request make sure that revoking a Task will not impact any users' pipelines.
```
If you mean "will not impact a single user's pipeline":
```suggestion
Before sending the pull request make sure that revoking a Task will not impact any user's pipeline.
```
Personally I would use the first one since it's more general

Author: Done

Author: I think its a technical document and we will have some user facing document from the docs team. So we does not have to be perfect as per the grammatical rules.

> ## Partner tekton task submission process

Author: Done.

> Partners can submit their tekton task in [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository by sending a PR.

Author: Done

> Build-definitions repository CI job will validate the PR against a [set of checks](#checks).

Author: Done.

Build-definitions repository maintainers needs to approve and merge the PR before the task can be used.

### How to submit a task

Create your task yaml and put it inside a directory of structure `partners/<task_name>/<task_version>/<task_name>.yaml` and 
also include a `OWNERS` file inside the task directory `partners/<task_name>/OWNERS`

> For example, if your task name is `my_task` and version of the your task is `0.1`, your task yaml and OWNERS file location will be `partners/my_task/0.1/my_task.yaml` and `partners/my_task/OWNERS`

Reviewer: ```suggestion
For example, if your task name is `my_task` and the version of your task is `0.1`, your task yaml and OWNERS file location will be `partners/my_task/0.1/my_task.yaml` and `partners/my_task/OWNERS`
```

Author: Done.

Reviewer: I believe there is still a redundant `the` left accidentally.
In "...the version of **the** your Task..."

Author: Good spot. Fixed now.

Send a PR to the [build-definitions](https://github.com/redhat-appstudio/build-definitions) repository containing your task yaml and OWNERS file.

### Checks
Task is validated against the following mandatory checks:
* Task follows the correct directory structure and includes OWNERS file (This check is a prerequisites for other checks to be continued)
* Task is not using any privilege escalations

> * Task schema is a valid one i.e. the task resource can be applied in the openshift cluster

Reviewer: I would think about changing `Task` to `The task`

Author: I would prefer to keep `Task` here.

## FAQs

### Can I validate the task locally?

> Yes you can also validate your task by running the script `hack/check-partner-tasks.sh` in your local clone. 

Reviewer: ```suggestion
Yes, you can also validate your task by running the script `hack/check-partner-tasks.sh` in your local clone. 
```

Author: Done.

> Make sure you have `oc` client installed and access to a openshift cluster before running the script.

Reviewer: ```suggestion
Make sure you have the `oc` client installed and access to an openshift cluster before running the script.
```

Author: Done.

> Make sure you have `oc` client installed and access to a openshift cluster before running the script.

Author: Added.

### Where to look for CI failures in my submitted task?

> Check the logs of `check-partner-tasks` step in the `build-definitions-pull-request` PR check.

Reviewer: ```suggestion
Check the logs of the `check-partner-tasks` step in the `build-definitions-pull-request` PR check.
```

Author: Done

> If you see step `yaml-lint-check` is failed, then your task yaml contains yaml-lint errors.

Reviewer: ```suggestion
If you see the step `yaml-lint-check` has failed, then your task yaml contains yaml-lint errors.
```

Author: Done.

> If you see step `yaml-lint-check` is failed, then your task yaml contains yaml-lint errors.

Author: Done.

### How to wire the task with the users pipeline?

Modify the PipelineRun to include the task yaml file path in the list of PipelineRun annotations `pipelinesascode.tekton.dev/task` 
and refer the task using task name in the `taskRef` section in the pipeline.

### Can I revoke a task?

Incase you have decided to revoke your task, you can also submit a PR removing the task directory from `partners` directory.

> Build-definitions repository maintainers will review and merge the PR.

Author: When I mentioned `you` I am not assuming `task user`, but mean to say `partner task` submitter. Will make it clear in the sentence. As a maintainer of the build-definitions repository, when we will be reviewing the PR. We need some mechanism to validate it a legit user. Submission request may come `partnerUser1` and deletion request maybe from `partnerUser` from the same partner group.

> Build-definitions repository maintainers will review and merge the PR.

Author: @tkdchen Agree with you. It requires further discussion.

## Support

> Please reach out to `#rhtap-users` channel in slack.

Reviewer: ```suggestion
Please reach out to the `#rhtap-users` channel in slack.
```

Author: Not applicable now.

> Please reach out to `#rhtap-users` channel in slack.

Author: As disscussed, will remove the `Support` part from this doc.
